api/v1beta1/clusterclass_types.go

```diff
@@ -209,6 +209,9 @@ type JSONSchemaProps struct {
 
 	// Default is the default value of the variable.
 	// +optional
+	// NOTE: Nullable variables are always defaulted to null independent of the default value.
+	// This is due to a limitation of the defaulting logic, which cannot differentiate
+	// between an unset variable and a variable set to null.
 	Default *apiextensionsv1.JSON `json:"default,omitempty"`
```

Comment:
It seems we're just documenting the limitation here, although I'm wondering why we can't differentiate between `null` as the json parameter and a nil default field?

Reply:
The problem is that I have to pass in the variable into the defaulting lib. I have to set something for the value which can be defaulted so I'm using `nil`. In case of a nullable field the defaulting lib doesn't default nil to something else because it cannot differentiate between the value `nil` and an unset value which should be defaulted.

The CRD behavior is that nullable fields which are set to null are not defaulted (as null is a valid value).


I could change the way we call the upstream defaulting logic that nullable variables are not defaulted at all. (defaulting currently means that a variable which does not exist on the Cluster resource is added with their default value)

Reply:
Hm, let's /hold this PR for now, I'll revisit the problem and maybe there's a way to solve this that the defaulting workings in a way that users would expect it to (and also is consistent with how we would default complex objects and CRDs work)

Reply:
I think I was slightly confused yesterday from too much debugging (there were a few other things I thought were broken :)) and going back and forth with different theories to realize that we can just fix that limitation. I'll resolve this conversation and open a new one.